server/listener.go

```diff
@@ -132,10 +135,32 @@ func (l *Listener) Close() error {
 		}
 	}
 	err = l.eg.Wait()
-	close(l.conns)
+	if !connResIsClosed(l.conns) {
+		close(l.conns)
+	}
 	return err
 }
 
 func (l *Listener) Addr() net.Addr {
 	return l.netListener.Addr()
 }
+
+func shutdownIsClosed(ch <-chan struct{}) bool {
```

Comment:
I don't think there's an official / safe way to check if a channel is closed, and my gut says there is a simpler way to do this... 

Is it possible to `nil` out `l.shutdown` and `l.conns` when `server.Close()` is called and then just check for `nil` in the `Close()` method instead? I think that would be a little bit cleaner and safer.